src/modules/commandrunner/src/lib/CommandRunner.cpp

```diff
@@ -1,465 +1,693 @@
 
-#include <ctime>
-#include <cstdio>
 #include <fstream>
-#include <Mmi.h>
-#include <sys/stat.h>
+#include <rapidjson/document.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+#include <rapidjson/prettywriter.h>
 
+#include <Command.h>
 #include <CommandRunner.h>
-#include <CommonUtils.h>
-
-const char commandStatusTemplate[] = "{\"CommandId\":\"\",\"ResultCode\":,\"TextResult\":\"\",\"CurrentState\":}";
-const char resultCodeChar[] = "-32767";
-
-static const std::string TMP_ENV_VAR = "TMPDIR";
-static const std::string TMP_DEFAULT_DIR = "/tmp";
-static const std::string TMP_FILE_PREFIX = "~osconfig-";
-
-OSCONFIG_LOG_HANDLE CommandRunnerLog::m_log = nullptr;
+#include <Mmi.h>
 
-CommandRunner::CommandRunner(std::string name, std::function<int()> persistentCacheFunction, unsigned int maxSizeInBytes) : curIndexCommandBuffer(0)
+constexpr const char info[] = R""""({
+    "Name": "CommandRunner",
+    "Description": "Provides functionality to remotely run commands on the device",
+    "Manufacturer": "Microsoft",
+    "VersionMajor": 2,
+    "VersionMinor": 0,
+    "VersionInfo": "Nickel",
+    "Components": ["CommandRunner"],
+    "Lifetime": 1,
+    "UserAccount": 0})"""";
+
+const std::string CommandRunner::PERSISTED_COMMANDSTATUS_FILE = "/etc/osconfig/osconfig_commandrunner.cache";
+
+std::mutex CommandRunner::s_diskCacheMutex;
+
+CommandRunner::CommandRunner(std::string clientName, unsigned int maxPayloadSizeBytes, bool usePersistedCache) :
+    m_clientName(clientName),
+    m_maxPayloadSizeBytes(maxPayloadSizeBytes),
+    m_usePersistedCache(usePersistedCache)
 {
-    clientName = name;
-    cacheFunction = persistentCacheFunction;
-    maxPayloadSizeInBytes = maxSizeInBytes;
+    if (m_usePersistedCache && (0 != LoadPersistedCommandStatus(clientName)))
+    {
+        OsConfigLogError(CommandRunnerLog::Get(), "Failed to load persisted command status for client %s", clientName.c_str());
+    }
+
+    // Start the worker thread
+    m_workerThread = std::thread(&CommandRunner::WorkerThread, std::ref(*this));
 }
 
 CommandRunner::~CommandRunner()
 {
-    OsConfigLogInfo(CommandRunnerLog::Get(), "CommandRunner shutting down");
+    OsConfigLogInfo(CommandRunnerLog::Get(), "CommandRunner %s shutting down", m_clientName.c_str());
+
+    while (!m_commandQueue.Empty())
+    {
+        std::shared_ptr<Command> command = m_commandQueue.Pop().lock();
+        if (nullptr != command)
+        {
+            command->Cancel();
+        }
+    }
+
+    // Push nullptr to signal the worker thread to exit
+    m_commandQueue.Push(std::weak_ptr<Command>());
 
     try
     {
-        commandWorkerThread.join();
+        if (m_workerThread.joinable())
+        {
+            m_workerThread.join();
+        }
+    }
+    catch (const std::exception& e)
+    {
+        OsConfigLogError(CommandRunnerLog::Get(), "Exception thrown while joining worker thread: %s", e.what());

```

Comment:
>OsConfigLogError(CommandRunnerLog::Get(), "Exception thrown while joining worker thread: %s", e.what()); [](http://example.com/codeflow?start=8&length=104)

This and one on line 75 seem strange to ne logged from within a class destructor. 

Edit the logged traces to say what happens overall because of this error.